src/main/java/ann/parser/Parser.java

```diff
@@ -0,0 +1,201 @@
+package ann.parser;
+
+import ann.commands.*;
+import ann.data.InputPattern;
+import ann.data.tasks.TaskType;
+
+/**
+ * Represents a parser to parse user input into appropriate commands.
+ *
+ * @author Hong Anh
+ * @version 0.1
+ */
+public class Parser {
+
+    /**
+     * Returns a Command which is the result of parsing the user input.
+     *
+     * @param input a String which is the user input.
+     * @return a Command.
+     */
+    public static Command parse(String input) {
+        if(input.toLowerCase().equals("bye")) {
+            return new ExitCommand();
+        } else if (input.toLowerCase().equals("list")){
+            return new ListCommand();
+        } else if (input.length() < 3) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 3).toLowerCase().equals("add")) {
+            return handleAdd(input.substring(3));
+        } else if(input.length() < 4) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 4).toLowerCase().equals("find")) {
+            return handleFind(input.substring(4));
+        } else if(input.substring(0, 4).toLowerCase().equals("mark")) {
+            return handleMark(input.substring(4));
+        } else if(input.length() < 6) {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        } else if(input.substring(0, 6).toLowerCase().equals("unmark")) {
+            return handleUnmark(input.substring(6));
+        } else if(input.substring(0, 6).toLowerCase().equals("delete")) {
+            return handleDelete(input.substring(6));
+        } else {
+            return new IncorrectCommand("Oops! Please enter a valid command!");
+        }
```

Comment:
```java
if (input.length() < 3) {
    // Check commands where 0 <= length < 3
} else if (input.length() < 4) {
    // Check commands where 3 <= length < 4
}
.
.
.
```

Do you think this portion will be more readable if it was changed to the above format? Personally I believe this block of code is not very straightforward as it seems some clauses depend on a certain length of input. Using the length as if-else conditions may help reinforce this idea.

Reply:
Actually for this, it's not just the length I have to check, but also whether it matches the command pattern, e.g. it can be shorter than 4 characters but if it is "bye" then I have to accept the command. I have to check the length in between as substring() would throw an index out of bounds exception otherwise. I'm still thinking about how this could be improved 😄